### PR TITLE
feat: Added Bead Sort in C++

### LINKED ISSUE #2072: Bead Sort
## 🚀 Feature
[DWOC Contribution]
Bead Sort in C++

(A clear and concise description of what the feature is.)

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## Motivation

Bead Sort (also known as Gravity Sort) was inspired from natural phenomenons and was designed keeping in mind objects(or beads) falling under the influence of gravity.

### PR DESCRIPTION
**Pull request by m-code12**
### Fixes
#2072 

### Have you read the [Contributing Guidelines on Pull Requests]
Yes

### Description

(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

### CODE REVIEW

**Comment by m-code12** [comment]
@HarshCasper @atarax665 Please review this pr.

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    for(auto const& i: nums)
+        cout<<i<<" ";
+    cout<<"\n";
+
+    auto max = *max_element(nums.begin(), nums.end());      // find maximum element in vector
+
+    vector<unsigned char> beads(max * nums.size(), 0);       // declare beads vector for sorting
+
+    for(auto i = 0; i < nums.size(); i++)                   // initialize the beads accordingly
+        for(auto j = 0; j < nums[i]; j++)
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Push all your core code logic into another function. Keep the main() function solely for the driver code.

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+Time Complexities:
+Bead sort can be implemented with four general levels of complexity, among others:
+O(1): The beads are all moved simultaneously in the same time unit, as would be the case with the simple physical example above.
+This is an abstract complexity, and cannot be implemented in practice.
+O(√N): In a realistic physical model that uses gravity, the time it takes to let the beads fall is proportional to the
+square root of the maximum height, which is proportional to n.
+O(n): The beads are moved one row at a time. This is the case used in the analog and digital hardware solutions.
+O(S), where S is the sum of the integers in the input set: Each bead is moved individually. This is the case when
+bead sort is implemented without a mechanism to assist in finding empty spaces below the beads, such as in software implementations.
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: No need to explain them. Just mention all the complexities plainly.

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/*Bead sort, also called as gravity sort, is a natural sorting algorithm since this algorithm was inspired from natural phenomenons
+and was designed keeping in mind objects(or beads) falling under the influence of gravity.The bead sort operation can be compared
+to the manner in which beads slide on parallel poles, such as on an abacus. However, each pole may have a distinct number of beads.*/
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Make it more compact and justify the formatting. It should be easy to read and consume

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+using namespace std;               //For cin and cout
+
+int main()
+{
+    vector<int> nums;                           // initialize a vector of integers
+    int a,n;                                    // Declaration of required variables
+    cout<<"Enter the number of elements :";
+    cin>>n;
+    cout<<"Enter the elements to be sorted : ";
+    for(int i=0;i<n;i++){                       //For loop for taking elements from the user, can be space-seperated as well
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        cin>>a;
+        nums.push_back(a);
+    }
+
+    nums.reserve(n);                        // reserve static vector for optimal performance gain
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    cout << "\nBefore sorting" << endl;    // Print vector before sorting
+    for(auto const& i: nums)
+        cout<<i<<" ";
+    cout<<"\n";
+
+    auto max = *max_element(nums.begin(), nums.end());      // find maximum element in vector
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    vector<unsigned char> beads(max * nums.size(), 0);       // declare beads vector for sorting
+
+    for(auto i = 0; i < nums.size(); i++)                   // initialize the beads accordingly
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            beads[i * max + j] = 1;
+    }
+
+    for(auto i = 0; i < nums.size(); i++)           // put sorted beads back into vector
+    {
+        for(auto j = 0; j < max && beads[i * max + j]; ++j)
+            nums[i] = j+1;
+    }
+
+    cout << "After sorting" << endl;               //For loop for printing out the array after sorting
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    for(auto const& i: nums)
+        cout << i << " ";
+    cout << endl;
+
+    nums.shrink_to_fit();               // Free memory space of vector
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
+/*Bead sort, also called as gravity sort, is a natural sorting algorithm since this algorithm was inspired from natural phenomenons
+and was designed keeping in mind objects(or beads) falling under the influence of gravity.The bead sort operation can be compared
+to the manner in which beads slide on parallel poles, such as on an abacus. However, each pole may have a distinct number of beads.*/
+
+#include <bits/stdc++.h>            // Header file to include STL library
+using namespace std;               //For cin and cout
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+int main()
+{
+    vector<int> nums;                           // initialize a vector of integers
+    int a,n;                                    // Declaration of required variables
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp]: Keep the Comments above the Line of Code

File: C-Plus-Plus/sort/Bead_Sort.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+Bead sort can be implemented with four general levels of complexity, among others:
+O(1): The beads are all moved simultaneously in the same time unit, as would be the case with the simple physical example above.
+This is an abstract complexity, and cannot be implemented in practice.
+O(√N): In a realistic physical model that uses gravity, the time it takes to let the beads fall is proportional to the
+square root of the maximum height, which is proportional to n.
+O(n): The beads are moved one row at a time. This is the case used in the analog and digital hardware solutions.
+O(S), where S is the sum of the integers in the input set: Each bead is moved individually. This is the case when
+bead sort is implemented without a mechanism to assist in finding empty spaces below the beads, such as in software implementations.
+*/
+
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/Bead_Sort.cpp:100]: Remove the extra line

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by m-code12** [comment]
@HarshCasper I have made all the changes, please review and merge it.

**Comment by m-code12** [comment]
@atarax665 Can you review and merge this?

**Comment by HarshCasper** [comment]
@m-code12 Make the README updates as well

**Comment by m-code12** [comment]
> @m-code12 Make the README updates as well

@HarshCasper Done

**Review by HarshCasper — APPROVED** [review]

**Comment by m-code12** [comment]
@HarshCasper Please could you add the relevant labels

**Pull request merged**